Move as_view/increment_version to its separate key.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53343 Implement public API InferenceMode and its error handling
* **#53342 Move as_view/increment_version to its separate key.**

Differential Revision: [D26973913](https://our.internmc.facebook.com/intern/diff/D26973913)

Commit:
Move as_view/increment_version to its separate key.

[ghstack-poisoned]

Commit:
Update on "Move as_view/increment_version to its separate key."




[ghstack-poisoned]

Commit:
Update on "Move as_view/increment_version to its separate key."




[ghstack-poisoned]

Commit:
Update on "Move as_view/increment_version to its separate key."




[ghstack-poisoned]

Commit:
Update on "Move as_view/increment_version to its separate key."




[ghstack-poisoned]

Commit:
Update on "Move as_view/increment_version to its separate key."




[ghstack-poisoned]

Commit:
Update on "Move as_view/increment_version to its separate key."




[ghstack-poisoned]

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+        ))
+        for r in cpp.return_names(f):
+            inplace_view_body.append(f'torch::autograd::increment_version({r});')
+    else:  # view op
```
There are ops that are both view and inplace (`transpose_()` for example).

ezyang replied:
I haven't the foggiest what `transpose_` is supposed to do. Does it restride the tensor? In that case, wouldn't you NOT version counter bump?

albanD replied:
You do want to version counter bump it. Because if an op saved the old version, it should fail as reading from the new Tensor (with updated strides) will return wrong result.

Note though that they are not properly handled today in the case of xla or cross-dtype view. But that will be covered in a separate issue: https://github.com/pytorch/pytorch/issues/49825#issuecomment-794466627
And this one will just aim at preserving the current behavior.

ezyang replied:
Variable saving does a shallow copy:

```
Tensor VariableHooks::tensor_data(const Tensor& self) const {
  TORCH_CHECK(self.defined(), "cannot call tensor_data() on undefined tensor");
  auto self_impl_copy = self.unsafeGetTensorImpl()->shallow_copy_and_detach(
    /*version_counter=*/self.unsafeGetTensorImpl()->version_counter(),
    /*allow_tensor_metadata_change=*/self.unsafeGetTensorImpl()->allow_tensor_metadata_change());
  return at::Tensor(self_impl_copy);
} 
```

Therefore stride update will not be seen by the saved copy. https://www.youtube.com/watch?v=5-1ar64-Rr4

bhosmer replied:
So just to confirm, this means that any metadata-only op, inplace or not, can be treated as functional w.r.t. versioning, right? i.e. if backwards always reads these shallow copies

ezyang replied:
Yes, because what version counter bump denotes (modulo detach) is a change to the *storage*; anything that isn't on storage (e.g., sizes, strides) doesn't count.

Commit:
Update on "Move as_view/increment_version to its separate key."




[ghstack-poisoned]

ezyang commented:
🕵️ You got any... generated code samples? 🕵️

ailzhang commented:
> 🕵️ You got any... generated code samples? 🕵️

Yup! here's the generated file https://gist.github.com/ailzhang/1d6a0f0a9d4a6970722ab8f2068e4c2a

ezyang commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+    if sig_group.faithful_signature is not None:
+        api_name = sig_group.faithful_signature.name()
+    else:
+        api_name = sig_group.signature.name()
```
@bdhirsh This seems to amplify the importance of https://github.com/pytorch/pytorch/issues/53023

ezyang commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+            unpacked_args=redispatch_args,
+        ))
+        for r in cpp.return_names(f):
+            inplace_view_body.append(f'torch::autograd::increment_version({r});')
```
This is unsound for functions which mutate input arguments but don't return them. We even have a list of these:

```
        if self.name.name.inplace:
            # TODO: fixme
            if str(self.name) not in [
                    '_amp_foreach_non_finite_check_and_unscale_',
                    '_foreach_add_.ScalarList',
                    '_foreach_sub_.ScalarList',
                    '_foreach_mul_.ScalarList',
                    '_foreach_div_.ScalarList',
                    '_foreach_add_.Scalar',
                    '_foreach_sub_.Scalar',
                    '_foreach_mul_.Scalar',
                    '_foreach_div_.Scalar',
                    '_foreach_add_.List',
                    '_foreach_sub_.List',
                    '_foreach_mul_.List',
                    '_foreach_div_.List',
                    '_foreach_exp_',
                    '_foreach_sqrt_',
                    '_foreach_abs_',
                    '_foreach_acos_',
                    '_foreach_asin_',
                    '_foreach_atan_',
                    '_foreach_ceil_',
                    '_foreach_cos_',
                    '_foreach_cosh_',
                    '_foreach_erf_',
                    '_foreach_erfc_',
                    '_foreach_expm1_',
                    '_foreach_floor_',
                    '_foreach_log_',
                    '_foreach_log10_',
                    '_foreach_log1p_',
                    '_foreach_log2_',
                    '_foreach_neg_',
                    '_foreach_tan_',
                    '_foreach_tanh_',
                    '_foreach_sin_',
                    '_foreach_sinh_',
                    '_foreach_round_',
                    '_foreach_lgamma_',
                    '_foreach_frac_',
                    '_foreach_reciprocal_',
                    '_foreach_sigmoid_',
                    '_foreach_trunc_',
                    '_foreach_addcmul_.Scalar',
                    '_foreach_addcdiv_.Scalar',
                    '_foreach_addcmul_.ScalarList',
                    '_foreach_addcdiv_.ScalarList',
                    '_foreach_zero_']:
                assert len(self.returns) == 1
```

ezyang replied:
But I guess this logic was taken from the preexisting autograd codegen, so it is technically no worse. Should file an issue for this.

albanD replied:
I think that none of these support autograd.
So VariableType should just be a no-op.

In this case, I don't think you need to generate a kernel for them because right now, you generate something pretty empty: https://ezyang.github.io/pytorch-ci-hud/build1/pytorch-master

ezyang replied:
If we ever want to differentiate through optimizer updates, the foreach kernels ought to support autograd eventually.

ailzhang replied:
Good point! I excluded these ops in InplaceOrView kernels and created an issue in https://github.com/pytorch/pytorch/issues/53796.

ezyang commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+        ))
+        for r in cpp.return_names(f):
+            inplace_view_body.append(f'torch::autograd::increment_version({r});')
+    else:  # view op
```
I think I'd probably appreciate an assert here that it actually is a view op.

ezyang commented:
I didn't compare the before and after code with a fine magnifying glass to make sure the movement was semantics preserving (Alban points out one case where it might not have been), but the general concept here seems fine. It sounds like there is some wrongness in the code though that may warrant some further work later.

ezyang commented:
(I'm going to leave final stamping to @albanD )

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+    dispatcher_sig = DispatcherSignature.from_schema(f.func)
+    dispatcher_exprs = dispatcher_sig.exprs()
+
+    ret_and_arg_types = ', '.join([dispatcher_sig.returns_type()] + [a.type.cpp_type() for a in dispatcher_exprs])
```
What is this used for?

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+    ret_and_arg_types = ', '.join([dispatcher_sig.returns_type()] + [a.type.cpp_type() for a in dispatcher_exprs])
+    # code-generated tracing kernels plumb and recompute dispatch keys directly through the kernel for performance.
+    # See Note [Plumbing Keys Through The Dispatcher] for details.
+    dispatch_key_set = 'ks & c10::DispatchKeySet(c10::DispatchKeySet::FULL_AFTER, c10::DispatchKey::InplaceOrView)'
```
The one used by VariableType is `ks & c10::after_autograd_keyset`
Why do you need something more complex here than `ks & c10::after_autograd_keyset & !c10::DispatchKey::InplaceOrView` ? Also what is the `FULL_AFTER` key?

albanD replied:
Ho.
That FULL_AFTER is a bit weird, since it's a keyset I was expecting it to be a key... ok

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+        api_name = sig_group.faithful_signature.name()
+    else:
+        api_name = sig_group.signature.name()
+    # inplace_view_body.append(THROW_IF_VARIABLETYPE_ON)
```
What is this comment?

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -247,22 +285,139 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+        type_definition_body=emit_inplace_or_view_body(fn),
+    )
+
+def modifies_arguments(f: NativeFunction) -> bool:
```
nit: function ordering. This should be before the functions that are above and below I think to be consistent?

albanD commented on tools/autograd/templates/InplaceOrViewType.cpp:
```diff
@@ -0,0 +1,32 @@
+
+#include <torch/library.h>
+
+#include "torch/csrc/autograd/function.h"
```
Is this and the Quantizer.h below actually needed?

albanD reviewed (commented):
lgtm.
As mentionned by Ed, some optimization can be done for ops that are inplace but not differentiable.

Commit:
Update on "Move as_view/increment_version to its separate key."




[ghstack-poisoned]

Commit:
Update on "Move as_view/increment_version to its separate key."




[ghstack-poisoned]

albanD commented on c10/core/DispatchKeySet.h:
```diff
@@ -203,6 +203,9 @@ constexpr DispatchKeySet autograd_dispatch_keyset = DispatchKeySet({
     DispatchKey::AutogradOther,
 });
 
+constexpr DispatchKeySet autograd_inplace_dispatch_keyset =
```
My understanding for the splitted AutogradCPU, AutogradCUDA, etc keys was that each backend can register a custom VariableType kernel if they want to.
So I guess that some backends are using that feature?

And it's these backends I mention above.

ailzhang replied:
Oh right, for AutogradXLA it's not a big problem since we only customize 2 operators using it for now. It might be for AutogradMLC tho, depends on how they want to support view/inplace...

albanD commented on c10/core/DispatchKeySet.h:
```diff
@@ -234,6 +237,12 @@ constexpr DispatchKeySet after_autograd_keyset = DispatchKeySet(
         c10::DispatchKey::AutogradOther
 );
 
+// The set of dispatch keys that com after InplaceOrView
```
nit: come

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -270,3 +312,117 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+    filtered_fns_with_infos = list(filter(use_derived, fns_with_infos))
+
+    fm.write_with_template('InplaceOrViewType%s.cpp' % suffix, 'InplaceOrViewType.cpp', lambda: {
+        'generated_comment': '@' + f'generated from {fm.template_dir}/InplaceOrViewType.cpp',
```
nit: why is the @ outside of the f string here?

albanD reviewed (approved):
LGTM

Commit:
Update on "Move as_view/increment_version to its separate key."


Differential Revision: [D26973913](https://our.internmc.facebook.com/intern/diff/D26973913)

[ghstack-poisoned]

Commit:
Update on "Move as_view/increment_version to its separate key."


Differential Revision: [D26973913](https://our.internmc.facebook.com/intern/diff/D26973913)

[ghstack-poisoned]

bhosmer commented on tools/autograd/gen_autograd.py:
```diff
@@ -55,12 +55,16 @@ def gen_autograd(
         # operator filter not applied as tracing sources are excluded in selective build
         gen_trace_type(out, native_functions_path, template_path)
 
+        gen_inplace_or_view_type(out, native_functions_path, fns_with_diff_infos, template_path)
```
this is very minor, but if you're back in here I'd suggest putting this right after the call to `gen_variable_type()`, just to visually reinforce the fact that they're related.

bhosmer commented on tools/autograd/gen_autograd.py:
```diff
@@ -55,12 +55,16 @@ def gen_autograd(
+
+
     # Generate Functions.h/cpp
     from .gen_autograd_functions import gen_autograd_functions_lib
```
not your change and no need to fix it here but time for some curmudgeonly grumbling about imports in the middle of code

bhosmer commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -99,6 +106,38 @@
+
+INPLACE_REDISPATCH = CodeTemplate("""\
+{
+  at::AutoDispatchBelowInplaceOrView guard(true);
```
Even if you decide not to actually remove the ctor param, I'd suggest we not supply it explicitly at callsites, again to clarify what the guard actually does.

bhosmer commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -270,3 +312,117 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
+    dispatcher_sig = DispatcherSignature.from_schema(f.func)
+    dispatcher_exprs = dispatcher_sig.exprs()
+
+    # code-generated tracing kernels plumb and recompute dispatch keys directly through the kernel for performance.
```
"tracing kernels" :P

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -645,14 +620,19 @@ def emit_call(f: NativeFunction, unpacked_bindings: List[Binding]) -> str:
         base_type_call = emit_dispatch_call(f, 'self_', unpacked_args)
 
+        if get_view_info(fn) is not None or modifies_arguments(f):
+            guard = 'at::AutoNonVariableTypeMode non_var_type_mode(true);'
```
nit: I'd aim for consistency in the names, either by renaming this one `guard` or renaming the other one something analogous to `non_var_type_mode` if you don't wanna generate unnecessary codegen diff :D

bhosmer commented on tools/codegen/model.py:
```diff
@@ -468,6 +468,55 @@ def from_dict(d: Dict[SchemaKind, NativeFunction]) -> Optional['StructuredNative
         )
 
+def is_foreach_op(name: str) -> bool:
+    return str(name) in [
```
yikes should we make this a Set? (not your code I know)

bhosmer reviewed (approved):
Nice! There's one nontrivial suggestion here, about removing the vestigial constructor param in the new guard... the rest is nits 😁

Commit:
Update on "Move as_view/increment_version to its separate key."


Differential Revision: [D26973913](https://our.internmc.facebook.com/intern/diff/D26973913)

[ghstack-poisoned]

Commit:
Update on "Move as_view/increment_version to its separate key."


Differential Revision: [D26973913](https://our.internmc.facebook.com/intern/diff/D26973913)

[ghstack-poisoned]
